python/ml4ir/base/pipeline.py

```diff
@@ -396,19 +397,36 @@ def get_relevance_model(self, feature_layer_keys_to_fns={}) -> RelevanceModel:
             aux_metrics = self.get_metrics(self.aux_metrics_keys)
 
         # Define scorer
-        scorer: RelevanceScorer = RelevanceScorer(
-            feature_config=self.feature_config,
-            model_config=self.model_config,
-            interaction_model=interaction_model,
-            loss=loss,
-            aux_loss=aux_loss,
-            aux_loss_weight=self.args.aux_loss_weight,
-            aux_metrics=aux_metrics,
-            output_name=self.args.output_name,
-            logger=self.logger,
-            file_io=self.file_io,
-            logs_dir=self.logs_dir_local
-        )
+        if self.args.monte_carlo_inference_trials > 0:
+            scorer: RelevanceScorer = MonteCarloScorer(
+                feature_config=self.feature_config,
+                model_config=self.model_config,
+                interaction_model=interaction_model,
+                loss=loss,
+                aux_loss=aux_loss,
+                aux_loss_weight=self.args.aux_loss_weight,
+                aux_metrics=aux_metrics,
+                output_name=self.args.output_name,
+                logger=self.logger,
+                file_io=self.file_io,
+                logs_dir=self.logs_dir_local,
+                monte_carlo_inference_trials=self.args.monte_carlo_inference_trials
+            )
+        else:
+
+            scorer: RelevanceScorer = RelevanceScorer(
+                feature_config=self.feature_config,
+                model_config=self.model_config,
+                interaction_model=interaction_model,
+                loss=loss,
+                aux_loss=aux_loss,
+                aux_loss_weight=self.args.aux_loss_weight,
+                aux_metrics=aux_metrics,
+                output_name=self.args.output_name,
+                logger=self.logger,
+                file_io=self.file_io,
+                logs_dir=self.logs_dir_local
+            )
```

Comment:
@mohazahran you can move this to a scorer factory python module to keep this clean - https://www.geeksforgeeks.org/factory-method-for-designing-pattern/

If you move the `monte_carlo_inference_trials` into the model config itself, then you can reduce the code duplication quite a bit.

Reply:
You can have some kind of a `get_scorer` function under `model/scoring/` and use that everywhere in the code.